### DISCUSSION TITLE
AuthSessionMissingError: Auth session missing

### DISCUSSION THREAD

**Discussion by QuDaMyker** [supabase/supabase#26791] [Questions] [answered]
I make a feature ResetPassword in Flutter, when usser request reset password, a email sent to users contains a link to reset

```javascript
/// Code Flutter
Future<Either<FailModel, bool>> resetPassword({required String email}) async {
    try {
      await supabase.auth.resetPasswordForEmail(
        email,
      );
      return Right(true);
    } catch (e) {
      debugPrint('[AuthController][resetPassword]: ${e.toString()}');
      return Left(FailModel(message: e.toString()));
    }
  }
```

```javascript
// code in react web
const supabase = createClient(
  supabaseUrl,
  supabaseKey,
);

async function updatePassword(newPassword) {
  try {
    const { data, error } = await supabase.auth.updateUser({
      password: newPassword,
    });

    if (error) {
      console.error("Error updating password:", error);
      return { error };
    }

    console.log("Password update successful:", data);
    return { data };
  } catch (error) {
    console.error("Unexpected error:", error);
    return { error };
  }
}

export { updatePassword };

```
I meet a bug in picture, help me with this bug please
![image](https://github.com/supabase/supabase/assets/93357243/72c3ba3b-911c-4cf8-a888-71f85eeb6a7b)

**Comment by rajnikantwebdev** [comment]
Getting the same error, I am using supabase in Next Js as I login and and move to other routes, I lost the access of session somehow. It works for the first time only.

**Comment by jagnani73** [comment]
Facing the same error

**Comment by shadeemerhi** [comment]
Facing the same error!

**Comment by QuDaMyker** [comment]
Need some helps pleaseaa

**Reply by rajnikantwebdev** [reply]
Auth session missing means, there is no current user. Verify the login, check if there any access token in the cookies, can you share more about your login method.

**Reply by QuDaMyker** [reply]
I followed to document on Official Supabase to reset the password (Flutter), an email contain a link to reset password sent, user clicked this link to navigate to website custom to reset (certainly, don't have step login there)

**Reply by faithandjs** [reply]
I have the same error. There is access token in cookies

**Comment by dshukertjr** [comment, marked as answer]
supabase_flutter and @supabase/ssr both use the [PKCE flow](https://supabase.com/docs/guides/auth/sessions/pkce-flow) by default, which disallows the sign-in to be complete if the sign-in flow was started in one environment (device / browser etc), and the sign-in link was clicked in another environment. You can use the [implicit flow](https://supabase.com/docs/guides/auth/sessions/implicit-flow) if you want to start the sign-in flow in one environment and complete it in another. 

You an set what flow to use in the `createClient` or `Supabase.initialize` call.

**Comment by baktun95827** [comment]
I have the same error , I think I am gonna give up on supabase auth

**Comment by FocusCookie** [comment]
I had a similar issue today that the session was not forwarded to my middleware in nextjs. After a long day of debugging and searching I went into the supabase discord and I found this posting:

https://github.com/supabase/ssr/commit/7dc1837dc4aba870b32792586969316df980ce07

Try to use npm i @supabase/ssr@patched for version 4.0.1 this fixes the issue with password recovery

**Reply by anthonyverducci** [reply]
Thanks! I was having the same issue and upgrading to 4.0.1 did the trick for me.

**Reply by sgiovo** [reply]
It worked for me too! ❤️

**Reply by arnaudschlupp** [reply]
Nice catch!

**Reply by dbuhon** [reply]
Yesss, thank you, version 0.4.1 fixed it.

**Reply by travisvn** [reply]
Thank you for this! 

`npm i @supabase/ssr@patched` solved it for me (going from version 0.4.0 to 0.4.1)

(had issues with user password reset flow)

**Reply by adg29** [reply]
Wow, this tip you shared dug me out of a rabbit hole. Thank you!

**Reply by infinia-yzl** [reply]
Life saver. Would have lost more hours on this if not for this post. Thank you ❤️

**Reply by Pokeko** [reply]
Thank you so much. I hate when it's about something you can't think about :(

**Reply by AhamSammich** [reply]
Solved my issues! Thanks!

**Reply by Vri6500** [reply]
wow, thank you for saving me! I already lost 3 hours on this

**Reply by JiayiLiuCA** [reply]
Saved my day! Thank you!

**Comment by rozabond** [comment]
I had a similar issues and just solved it from reddit and other forum..

'<'h2'>Reset Password<'/h2'>'

'<'p'>Follow this link to reset the password for your user:<'/p'>'
'<'p'><'a' href="{{ .SiteURL }}?token_hash={{ .TokenHash }}&type=email&redirectUrl={{ .RedirectTo }}&email={{.Email}}&token={{.Token}}">Reset Password<'/a'><'/p'>'
'<'p>{{ .Token }}<'/p'>'

used this email template for reset password to get the token!.

Used the token to sign in to the supabase via upabase.auth.verifyOtp()

The rest work like magic :)

**Comment by RemyNtshaykolo** [comment]
I’ve spent a considerable amount of time troubleshooting this issue on React Native/Expo, and here’s the solution that finally worked for me. Basically this catch the supabase deeplink redirection and iniatiate a user session using the access token and refresh token included in the redirection url.
I found the solution here : https://blog.theodo.com/2023/03/supabase-reset-password-rn/

``` javascript
import * as Linking from "expo-linking";

Linking.addEventListener("url", async (event) => {
  const parseSupabaseUrl = (url: string) => {
    let parsedUrl = url;
    if (url.includes("#")) {
      parsedUrl = url.replace("#", "?");
    }
    return Linking.parse(parsedUrl);
  };
  const url = parseSupabaseUrl(event.url);

  console.log("parsed url", url);

  const { data: setSessionData, error: setSessionError } =
    await supabase.auth.setSession({
      access_token: url.queryParams.access_token,
      refresh_token: url.queryParams.refresh_token,
    });
  console.log("setSessionData", setSessionData);
  console.log("setSessionError", setSessionError);
});
```

**Reply by sapphire-haeward** [reply]
I'm using React Native and expo too. I can't find the access_token and refresh_token. 

Wait wait wait, the missing piece of the piece of the puzzle was the addEventListerner.
If you don't have that you won't get ANYTHING.

**Reply by JamieGoodson** [reply]
Also using Expo/React Native here for our mobile app.

In my case the user is emailed a password reset link that takes the user to a page on our website (not the mobile app) since this is what we want. Therefore I used your above method to extract the auth credentials from the URL to create the session **and then** we can reset the password, before asking them to return to our mobile app to sign in again. Your solution works beautifully 🙏

**Comment by gpiechnik2** [comment]
I also encountered the same error, but it is simple to solve. Case:

1. a user wants to reset his password, so he fills out a reset form.
2. an email address arrives with the password reset link
3. the user is redirected to a page where he enters the password

So far everything is well described here:

https://supabase.com/docs/guides/auth/passwords?queryGroups=language&language=js

Then there is an error with the session, because the documentation lacks information on how to handle it correctly. My working example:


/api/auth/new-password/route.ts

```typescript
import { NextResponse } from 'next/server'
import { createClient } from '@/utils/supabase/server'

export async function POST(request: Request) {
    const requestUrl = new URL(request.url)
    const formData = await request.formData()
    const password = String(formData.get('password'))
    const code = requestUrl.searchParams.get('code')

    if (code) {
        const supabase = createClient()
        await supabase.auth.exchangeCodeForSession(code)

        const { data, error } = await supabase.auth.updateUser({
            password: password
        })

        if (error) {
            return NextResponse.json({ error: error }, { status: 500 })
        }
    }

    return NextResponse.redirect(new URL('/dashboard/library', request.url), {
        status: 302,
    })
}

```

**Reply by wesleydmscn** [reply]
The `await supabase.auth.exchangeCodeForSession(code)` works for me, thank you! 😄

**Reply by raihan-ramadhan** [reply]
thanks bro, why they not include/explain this on reset-password doc 😒

**Comment by gpiechnik2** [comment]
Also, for some unknown for me reasons avoid using:

```typescript
const { data, error } = await supabase.auth.getUser()
```

instead do:

```typescript
const {
    data: { user },
  } = await supabase.auth.getUser();
```

I wrote "unknown", because I noticed that when I use the first way, I get the same error as in the thread (it seems to be some problem affecting different use cases)

**Reply by rajasekarpr13** [reply]
I too faced the same issue, basically I was checking if the user session exists in the navbar and it was throwing this error in all routes, when the user is not logged in.  After changing to the 2nd way there is no error.

**Comment by phillipo6** [comment]
In case anyone stumbles upon this, i had the same error. My problem was setting the siteurl in supabase to 127.0.0.1 and using localhost to access my nextjs frontend. fixed it once i was using the same hostname/ip

**Reply by ShaneK** [reply]
This helped me a ton, thank you!

**Reply by davisg123** [reply]
🙌 Thank you!

**Reply by jakfils** [reply]
I was losing my mind — your answer brought me back. Eternal thanks! 😅❤️

**Comment by CarrettaRiccardo** [comment]
For me it was a situation where i created a supabase `createServerClient` with `SERVICE_ROLE_KEY` (without user's cookies) and tried to execute a `supabase.auth.updateUser()` (which is tied to the user's cookies)

**Comment by bigyankarki** [comment]
I was working on a mobile app, and my reset password flow was in a site. So i just used this `const { data, error } = await this.supabase.auth.admin.updateUserById(userId, { password });` to reset the password in backend.

**Comment by skiunke** [comment]
I had a similar problem when I forgot to check if a session was available when calling the refresh JWT endpoint.

**Comment by bayramorhan** [comment]
for those who getting this error on development environment, you might need to set "secure: false" in your nuxt.config

```
 supabase: {
    redirect: false,
    cookieOptions: {
      secure: false,
    },
  },
```

**Reply by LiuShengYuan123** [reply]
Awesome！！！

**Comment by EtcetFelix** [comment]
Had the same issue and spent hours trying to fix it. 
The session was properly being authenticated on the api/auth/confirm page after clicking the link in the email, but when redirecting to the update-password page, the session was no longer present. Kinda hard to debug because it just looked like the email link didn't work, until I logged the session everywhere.

Our application followed what the docs said:
* Reset email template link: {{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash }}&type=recovery&next=/update-password
* Local email config (inside config.toml in supabase folder):
```
[auth.email.template.recovery]
subject = "Reset Your Password"
content_path = "./supabase/templates/recovery.html"
```
_Relevant links_: 
https://supabase.com/docs/guides/local-development/customizing-email-templates
https://supabase.com/docs/guides/auth/auth-email-templates
* confirm route code: 
```
export async function GET(request: NextRequest) {
  const { searchParams } = new URL(request.url)
  const token_hash = searchParams.get('token_hash')
  const type = searchParams.get('type') as EmailOtpType | null
  const next = searchParams.get('next') ?? '/dashboard'
  const token = searchParams.get('token') // For PKCE flow
  const code = searchParams.get('code') // Code from email link
  const redirectTo = request.nextUrl.clone()
  redirectTo.pathname = next
const supabase = await createClient()
if (token_hash && type) {
    const { error } = await supabase.auth.verifyOtp({
      type,
      token_hash,
    })
    
    if (!error) {
      console.log("OTP verification successful, redirecting to:", next);
      const response = NextResponse.redirect(redirectTo)
      return response
    }
    
    console.error("Error verifying OTP:", error);
    return NextResponse.redirect(new URL('/login?error=verification_failed', request.url))
  }
```

For all of the above, the flow worked well. Session was authenticated, but on the update-password page, the session didn't exist.
The cause for me was something to do with cookies not properly being propagated to the update-password page due to it being a server component. We're using next.js app router, which by default makes components server components unless you set 'use client' at the top. 
I verified this to be the reason by checking the cookies sent to and from the browser via the network tab request headers on browser dev tools. Server components don't inherently have access to cookies, so they rely on the browser sending and receiving them via headers. For some reason, the cookies weren't being sent when I had my page as a server component, not sure why. Maybe I wasn't doing something right.

Whatever the case, on the update-password page I switched to using 'use client' then used router.push instead of redirect. Fixed it for me. 

Hopefully this saves someone a couple hours.

**Comment by hnykda** [comment]
Just a note that this error is alsy simply raised in scenarios where it's valid that the session doesn't exist. E.g. I had to whitelist it for reporting purposes:

```
import { createClient } from "@cohort/trinket/lib/db/client/server";
import { LoginPage } from "./LoginPage";

// Duplicated in portal and deep-disposition, as it's non-trivial to share
// react components between the two projects (surely feasible though).
export default async function LoginPageServer({
  searchParams,
}: {
  searchParams: Promise<{ next?: string }>;
}) {
  const supabase = await createClient();

  const { data, error } = await supabase.auth.getUser();

  // AuthSessionMissingError is expected when the user is not logged in
  if (error && error.name !== "AuthSessionMissingError") {
    console.error("error while getting user", error);
    return <p>Sorry, something went wrong with the login</p>;
  }

  return <LoginPage user={data.user} nextUrl={(await searchParams).next} />;
}

```

**Comment by jeszdav** [comment]
Make sure you import the `createClient` method from supabase/server and not from  supabase/client like: `import { createClient } from '@/lib/supabase/server'`.  At least this is what solved it for me.
